messenger.go

```diff
@@ -916,6 +935,111 @@ func (p *postProcessor) matchMessage(message *protocol.Message, chats []*Chat) (
 	}
 }
 
+type EnsDetails struct {
+	Name      string `json:"name"`
+	PublicKey string `json:"publicKey"`
+}
+
+type EnsResponse struct {
+	Name            string           `json:"name"`
+	Verified        bool             `json:"verified"`
+	Error           error            `json:"error"`
+	PublicKey       *ecdsa.PublicKey `json:"-"`
+	PublicKeyString string           `json:"publicKey"`
+}
+
+func (m *Messenger) verifyEnsName(ensInfo EnsDetails, ethclient *ethclient.Client, output chan EnsResponse) {
+	publicKeyStr := ensInfo.PublicKey
+	ensName := ensInfo.Name
+	m.logger.Info("Resolving ENS name", zap.String("name", ensName), zap.String("publicKey", publicKeyStr))
+	response := EnsResponse{
+		Name:            ensName,
+		PublicKeyString: publicKeyStr,
+	}
+
+	expectedPubKeyBytes, err := hex.DecodeString(publicKeyStr)
+	if err != nil {
+		response.Error = err
+		output <- response
+		return
+	}
+
+	publicKey, err := crypto.UnmarshalPubkey(expectedPubKeyBytes)
+	if err != nil {
+		response.Error = err
+		output <- response
+		return
+	}
+
+	// Resolve ensName
+	resolver, err := ens.NewResolver(ethclient, ensName)
+	if err != nil {
+		m.logger.Error("error while creating ENS name resolver", zap.String("ensName", ensName), zap.Error(err))
+		response.Error = err
+		output <- response
+		return
+	}
+	x, y, err := resolver.PubKey()
+	if err != nil {
+		m.logger.Error("error while resolving public key from ENS name", zap.String("ensName", ensName), zap.Error(err))
+		response.Error = err
+		output <- response
+		return
+	}
+
+	// Assemble the bytes returned for the pubkey
+	pubKeyBytes := elliptic.Marshal(crypto.S256(), new(big.Int).SetBytes(x[:]), new(big.Int).SetBytes(y[:]))
+
+	response.PublicKey = publicKey
+	response.Verified = bytes.Equal(pubKeyBytes, expectedPubKeyBytes)
+	output <- response
+}
+
+// VerifyENSName verifies that a registered ENS name matches the expected public key
+func (m *Messenger) VerifyENSNames(ensDetails []EnsDetails, rpcEndpoint, contractAddress string) (map[string]EnsResponse, error) {
+	var contacts []Contact
+	ctx, cancel := context.WithTimeout(context.Background(), contractQueryTimeout)
+	defer cancel()
+
+	ch := make(chan EnsResponse)
+	response := make(map[string]EnsResponse)
+
+	ethclient, err := ethclient.DialContext(ctx, rpcEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ensInfo := range ensDetails {
+		go m.verifyEnsName(ensInfo, ethclient, ch)
```

Comment:
Alternatively:
```suggestion
		go func() { ch <- m.verifyEnsName(ensInfo, ethclient) }()
```
so that `verifyEnsName` signature looks simpler.

Reply:
yes, this is much better